k8s cni integration and related netplugin changes/fixes

jojimt opened the pull request.

jojimt commented on netmaster/master/endpointGroup.go:
```diff
@@ -17,7 +17,7 @@ package master
 import (
 	"errors"
-	"fmt"
+	//"fmt"
```
Looks like one commit is missing. Let me fix this.

jainvipin commented:
@jojimt - can the title of the PR be `k8s cni integration and related netplugin changes/fixes`

jojimt commented:
Corrected it (what was meant to be description appeared as title…)

From: Vipin Jain [mailto:notifications@github.com]
Sent: Friday, January 08, 2016 10:21 PM
To: contiv/netplugin
Cc: Joji Mekkattuparamban (joji)
Subject: Re: [netplugin] Kuberenetes integration. Summary of changes: 1) Added CNI exec plugin 2) Added k8s plugin to mgmnt function 3) Added command line arguments to enable kubernetes mode (default is docker) 4) Fixed a mutex issue in the CreateEndpoint path....


@jojimt<https://github.com/jojimt> - can the title of the PR be k8s cni integration and related netplugin changes/fixes

—
Reply to this email directly or view it on GitHub<https://github.com/contiv/netplugin/pull/224#issuecomment-170201199>.

jainvipin commented:
@jojimt: we also need to update the docs with k8 integration, either with this PR or separate one, how to use k8s with contiv, etc.

jojimt commented:
Yes, that will be part of the next (Vagrant) PR.

From: Vipin Jain [mailto:notifications@github.com]
Sent: Sunday, January 10, 2016 11:29 AM
To: contiv/netplugin
Cc: Joji Mekkattuparamban (joji)
Subject: Re: [netplugin] k8s cni integration and related netplugin changes/fixes (#224)


@jojimt<https://github.com/jojimt>: we also need to update the docs with k8 integration, either with this PR or separate one, how to use k8s with contiv, etc.

—
Reply to this email directly or view it on GitHub<https://github.com/contiv/netplugin/pull/224#issuecomment-170383800>.

shaleman commented on contivk8s/clients/network.go:
```diff
@@ -0,0 +1,122 @@
+/***
```
As i mentioned, can we move contivk8s into mgmtfn/k8splugin hierarchy?

jojimt replied:
Yes, will do that.

shaleman commented on netmaster/master/endpointGroup.go:
```diff
@@ -80,13 +80,15 @@ func CreateEndpointGroup(tenantName, networkName, groupName string, epgID int) e
 
 	// params for docker network
-	subnetCIDR := fmt.Sprintf("%s/%d", nwCfg.SubnetIP, nwCfg.SubnetLen)
+	if GetClusterMode() == "docker" {
```
Better way to do this might be to create mgmtfn abstraction here. so that we can move docker/swarm specific things into docker mgmtfn.
We can do this as a subsequent PR too.

jojimt replied:
I agree. However, I would rather do it as a separate PR.

shaleman commented on contivk8s/clients/network.go:
```diff
@@ -0,0 +1,122 @@
+)
+
+const (
+	nwURL = "http://localhost"
```
I'm not sure if we need this since we are using unix sockets

jojimt replied:
Actually we do. This is an artifact of the net/http package where it rejects the URL if it did not have a domain name.

shaleman commented on contivk8s/k8s_cni.go:
```diff
@@ -0,0 +1,129 @@
+	// Add to contiv network
+	result, err := nc.AddPod(pInfo)
+	if err != nil {
+		log.Fatalf("EP create failed -- %s", err)
```
How is error communicated back to kubelet? is it based on exit codes?
does log.Fatalf() set the correct exit code?

jojimt replied:
kubelet just checks for status 0 and treats everything else as an error without interpreting it.

shaleman commented on mgmtfn/k8splugin/driver.go:
```diff
@@ -0,0 +1,370 @@
+}
+
+// setIfAttrs sets the required attributes for the container interface
+func setIfAttrs(ifname, netns, cidr, newname string) error {
```
Here is how old crtclient code was moving container to namespace without using nester,
https://github.com/contiv/netplugin/blob/k8s-crt-pre-deprecate/crtclient/docker/docker.go#L122

jojimt replied:
Yes I see the netns lib. I am not sure it will do everything we need. But I can look at that separately.

shaleman commented on mgmtfn/k8splugin/driver.go:
```diff
@@ -0,0 +1,370 @@
+		return &resp, err
+	}
+
+	netw, _ := kubeAPIClient.GetPodLabel(pInfo.K8sNameSpace, pInfo.Name, "network")
```
error handling missing? :)

jojimt replied:
Actually, if the first check passes the subsequent ones are guaranteed to pass. I will add a comment to that effect.

shaleman commented on mgmtfn/cniapi/api.go:
```diff
@@ -0,0 +1,48 @@
+const PluginPath = "/run/contiv"
+
+// DriverName is the name of the listen socket for netplugin
+const DriverName = "netplugin"
```
Looks like this const is not being used?

jojimt replied:
Right, will remove.

shaleman commented on mgmtfn/k8splugin/cniserver.go:
```diff
@@ -0,0 +1,162 @@
+
+// contivKubeCfgFile holds credentials to access k8s api server
+const (
+	contivKubeCfgFile = "/opt/contiv/config/contiv.json"
```
I'm not too happy about requiring user to create this config file.
But, i don't have an alternative at this point. so, we can keep it for now.

jojimt replied:
For the common scenarios, this is created using ansible; so it's not that bad. But, I agree.

jainvipin replied:
Is this a security problem?

jojimt replied:
Well, kube set up uses these certs. Therefore, I don't consider it a security problem specific to contiv. There is also a different way of authenticating that I can look into, separately. However, I don't know that it is any more secure.

shaleman commented on mgmtfn/cniapi/api.go:
```diff
@@ -0,0 +1,48 @@
+/***
```
as discussed, lets move this under mgmtfn/k8splugin hierarchy too.

jojimt replied:
sure.

shaleman commented:
LGTM other than the minor comments

jainvipin commented on mgmtfn/k8splugin/contivk8s/clients/network.go:
```diff
@@ -0,0 +1,122 @@
+}
+
+// AddPod adds a pod to contiv using the cni api
+func (c *NWClient) AddPod(podInfo interface{}) (*cniapi.RspAddPod, error) {
```
Unit tests for AddPod and DelPod public functions?

jojimt replied:
Because we moved the kubeserver access inside netplugin, these functions will need access to a kube server. I think sys-test is the only option. Will add it after vagrant work is completed.

jainvipin replied:
There should be a way to mock k8s REST handler; for example someone had mentioned an easy way to mock if we made the correct REST calls. Ultimately if you see these functions changing, then making unit tests pays off...

jojimt replied:
I guess I can do that. Will add it.

jainvipin commented on mgmtfn/k8splugin/cniserver.go:
```diff
@@ -0,0 +1,162 @@
+
+// ContivConfig holds information passed via config file during cluster set up
+type ContivConfig struct {
+	K8sAPIServer string `json:"K8S_API_SERVER,omitempty"`
```
Do we specify which version of API that APIServer interaction is on with?

jojimt replied:
Right now it's hard coded to v1 like this:
const (
        nsURL = "/api/v1/namespaces/"
)
We can make it configurable, if necessary.

jainvipin replied:
Ah... as long it is specified; I don't think there is a need for it to be configurable.

jainvipin commented on mgmtfn/k8splugin/contivk8s/clients/network.go:
```diff
@@ -0,0 +1,122 @@
+	defer r.Body.Close()
+
+	switch {
+	case r.StatusCode == int(404):
```
This code in switch statement looks similar in AddPod and DelPod; shoudl we combine the error handling in a common place?

jojimt replied:
OK, will do.

jojimt replied:
After looking at this again, this does not fit as a separate routine as we have different return signatures. It is only 4 lines of code anyway.

jainvipin commented on mgmtfn/k8splugin/driver.go:
```diff
@@ -0,0 +1,371 @@
+}
+
+// nsToPID is a utility that extracts the PID from the netns
+func nsToPID(ns string) (int, error) {
```
Perhaps we can use this in mesos plugin too... should this belong in netplugin/utils?

jojimt replied:
As we don't know the exact needs of mesos plugin at this point, I will leave this to @unclejack to refactor and move as appropriate.

jainvipin commented on mgmtfn/k8splugin/driver.go:
```diff
@@ -0,0 +1,371 @@
+	err = epCleanUp(epReq)
+	resp.EndpointID = pInfo.InfraContainerID
+	return resp, err
+}
```
same comment - no unit tests for any of the functions in this file either...

jojimt replied:
I have added UT for the REST interface. This one, seems like it will need to be covered by system test as this module makes multiple external calls.

jainvipin commented on mgmtfn/k8splugin/kubeClient.go:
```diff
@@ -0,0 +1,174 @@
+
+	defer r.Body.Close()
+	switch {
+	case r.StatusCode == int(404):
```
repetitive code - was also there in AddPod and DelPod functions above...

jainvipin commented on mgmtfn/k8splugin/kubeClient.go:
```diff
@@ -0,0 +1,174 @@
+	p.setDefaults(ns, name)
+
+	meta := m.(map[string]interface{})
+	l, ok := meta["labels"]
```
minor: if statement below this and this statement can be combined

jojimt replied:
I have intentionally stayed away from that style because I have seen that it limits the scope of the boolean to just the if. I find that scoping confusing.

jainvipin commented on netmaster/master/netmaster.go:
```diff
@@ -36,6 +36,13 @@ const (
+	clusterMode string
+}
+
+var masterRTCfg nmRunTimeConf
```
Should this type of information be added in global runtime structure? This allows for netmaster restartability - @shaleman WDYT?

jojimt replied:
This is basically environment information. I don't think we should save it in etcd. If the netmaster is restarted, that would be still in the same environment.

jainvipin replied:
Hmm... got it; is this read always after netmaster start? If yes, we are covered.

jojimt replied:
It's a command line argument, expected to come from the service env file.

shaleman replied:
@jainvipin: my thinking is:
(1) Plugin modes (like docker/kubernetes/mesos plugins) that need to be specified before netplugin/netmaster initializes should be command line arguments.
(2) Cluster modes/Fabric modes/control plane modes should be an Intent that can change at runtime. We need little more infra in netplugin/netmaster for this to happen. Expect to work on this in next few PRs.

jainvipin commented:
@jojimt - great work! thanks!
please identify the changes we need to make before the checkin vs. tests that you want to address as a part of subsequent checkins.

jainvipin commented:
LGTM
